contentcuration/contentcuration/utils/recommendations.py

```diff
@@ -61,67 +76,307 @@ def create_backend(self) -> Backend:
+        :return: The cached response if it exists, otherwise None.
+        :rtype: Union[EmbeddingsResponse, None]
+        """
         try:
-            embed_topics_request = EmbedTopicsRequest(json=topics)
-            return self.backend.make_request(embed_topics_request)
+            request_hash = self._generate_request_hash(request)
+            override_threshold = self._extract_override_threshold(request)
+            data = list(RecommendationsCache.objects
+                        .filter(request_hash=request_hash, override_threshold=override_threshold)
+                        .order_by('override_threshold', 'rank')
+                        .values('contentnode_id', 'rank'))
+            if len(data) > 0:
+                return EmbeddingsResponse(data=data)
+            else:
+                return None
         except Exception as e:
-            return EmbeddingsResponse(error=e)
+            logging.exception(e)
+            return None
+
+    def _generate_request_hash(self, request) -> str:
+        """
+        Generates a unique hash for a given request. It serializes the request
+        attributes that make it unique, then generates a hash of this serialization.
+
+        To prevent cache duplication, the hash is generated
+        independent of the override_threshold parameter.
+
+        :param request: The request for which to generate a unique hash.
+        :return: A unique hash representing the request
+        """
 
-    def embed_content(self, nodes: List[ContentNode]) -> EmbeddingsResponse:
+        params_copy = request.params.copy() if request.params else {}
+        params_copy.pop('override_threshold', None)
 
+        unique_attributes = json.dumps({
+            'params': params_copy,
+            'json': request.json,
+        }, sort_keys=True).encode('utf-8')
+
+        return hashlib.md5(unique_attributes).hexdigest()
+
+    def cache_embeddings_request(self, request: BackendRequest, response: BackendResponse) -> bool:
+        """
+        Caches the recommendations request and response. It performs a bulk insert of the
+        recommendations into the RecommendationsCache table, ignoring any conflicts.
+
+        :param request: The request to cache.
+        :param response: The response to cache.
+        :return: A boolean indicating whether the caching was successful.
+        :rtype: bool
+        """
+
+        try:
+            nodes = self._extract_data(response)
+            request_hash = self._generate_request_hash(request)
+            existing_cache = set(RecommendationsCache.objects.filter(request_hash=request_hash)
+                                 .values_list('contentnode_id', flat=True))
+            override_threshold = self._extract_override_threshold(request)
+            new_cache = [
+                RecommendationsCache(
+                    request_hash=request_hash,
+                    contentnode_id=node['contentnode_id'],
```

Comment:
what happens if the curriculum automation API is out-of-date and returns a node that was deleted (doesn't exist and the FK would be invalid)?

Reply:
I think we need to check against the database to ensure that these nodes exist before attempting to cache, that way we ensure that whatever exists in the cache is usable data.

Reply:
That sounds like a good defensive check to implement